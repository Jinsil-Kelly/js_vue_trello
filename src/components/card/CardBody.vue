<template>
  <div>
    <h3>Description</h3>
    <textarea
      class="form-control"
      cols="30"
      rows="3"
      placeholder="Add a more detailed description..."
      :readonly="!toggleDesc"
      @click="$emit('onToggleDesc')"
      @blur="$emit('onBlurDesc', $event.target.value)"
      :value="description"
    ></textarea>
  </div>
</template>

<script>
export default {
  name: "EditCardHeader",
  props: ["description", "toggleDesc"]
};
</script>

<style scoped></style>
