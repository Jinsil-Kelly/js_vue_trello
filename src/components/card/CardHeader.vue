<template>
  <div class="modal-card-header">
    <div class="modal-card-header-title">
      <input
        class="form-control"
        :readonly="!toggleTitle"
        @click="$emit('onToggleTitle')"
        @blur="$emit('onBlurTitle', $event.target.value)"
        type="text"
        :value="title"
      />
    </div>
    <a class="modal-close-btn" href="" @click.prevent="$emit('close')"
      >&times;</a
    >
  </div>
</template>

<script>
export default {
  name: "EditCardHeader",
  props: ["title", "toggleTitle"]
};
</script>

<style scoped></style>
