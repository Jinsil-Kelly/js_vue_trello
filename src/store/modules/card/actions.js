import * as api from "@/api/api";

export default {
  //id is bId
  ADD_CARD({ dispatch }, { id, title, listId, pos }) {
    return api.card
      .create(title, listId, pos)
      .then(() => dispatch("board/FETCH_BOARD", { id }, { root: true }));
  },

  //id is cId
  FETCH_CARD({ commit }, { id }) {
    return api.card.fetch(id).then(data => {
      commit("SET_CARD", data.item);
    });
  },

  //id is cId
  UPDATE_CARD({ dispatch }, { id, title, description, pos, listId, bId }) {
    return api.card
      .update(id, { title, description, pos, listId })
      .then(() => dispatch("board/FETCH_BOARD", { id: bId }, { root: true }));
  }
};
