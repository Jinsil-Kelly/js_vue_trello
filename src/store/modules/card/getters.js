export default {
  card: state => state.card
};
