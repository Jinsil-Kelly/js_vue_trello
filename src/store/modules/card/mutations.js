export default {
  SET_CARD(state, card) {
    state.card = card;
  }
};
