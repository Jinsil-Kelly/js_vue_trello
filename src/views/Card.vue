<template>
  <Modal class="modal-card">
    <div slot="header">
      <EditCardHeader :title="card.title" />
    </div>
    <div slot="body">
      <h3>Description</h3>
      <textarea
        class="form-control"
        cols="30"
        rows="3"
        placeholder="Add a more detailed description..."
        :readonly="!toggleDesc"
        @click="toggleDesc = true"
        @blur="onBlurDesc"
        v-model="card.description"
        ref="inputDesc"
      ></textarea>
    </div>
    <div slot="footer"></div>
  </Modal>
</template>

<script>
import { createNamespacedHelpers, mapGetters } from "vuex";
const { mapActions } = createNamespacedHelpers("card");
import Modal from "@/components/Modal.vue";
import EditCardHeader from "@/components/EditCardHeader.vue";

export default {
  components: { Modal, EditCardHeader },
  data() {
    return {
      toggleTitle: false,
      toggleDesc: false
    };
  },
  computed: {
    ...mapGetters("board", ["board"]),
    ...mapGetters("card", ["card"])
  },
  created() {
    // this.fetch()
    this.FETCH_CARD({ id: this.$route.params.cId });
  },
  methods: {
    ...mapActions(["FETCH_CARD", "UPDATE_CARD"]),
    // fetch(){
    //   this.FETCH_CARD({ id:this.$route.params.cId });
    // },
    onClose() {
      this.$router.push(`/b/${this.board.id}`);
    },
    onBlurTitle() {
      this.toggleTitle = false;
      const title = this.$refs.inputTitle.value.trim();
      if (!title) return;
      this.UPDATE_CARD({
        id: this.card.id,
        title,
        bId: this.$route.params.bId
      }).then(() => this.FETCH_CARD({ id: this.$route.params.cId }));
    },
    onBlurDesc() {
      this.toggleDesc = false;
      const description = this.$refs.inputDesc.value.trim();
      this.UPDATE_CARD({
        id: this.card.id,
        description,
        bId: this.$route.params.bId
      });
    }
  }
};
</script>

<style>
.modal-card .modal-container {
  min-width: 300px;
  max-width: 800px;
  width: 60%;
}
.modal-card-header-title {
  padding-right: 30px;
}
.modal-close-btn {
  position: absolute;
  top: 0px;
  right: 0px;
  font-size: 24px;
  text-decoration: none;
}
.modal-card-header {
  position: relative;
}
</style>
