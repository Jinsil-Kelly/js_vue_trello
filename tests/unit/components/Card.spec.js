import Card from "@/views/Card.vue";
import { shallowMount, createLocalVue } from "@vue/test-utils";
import Vuex from "vuex";
import {fetchedCard} from "../fixtures/card";
import cardBoard from "../fixtures/cardBoard";
const localVue = createLocalVue();
localVue.use(Vuex);
describe("Card", () => {
  let wrapper;
  const actions = {
    FETCH_CARD: jest.fn(),
    UPDATE_CARD: jest.fn()
  };

  const $route = {
    params: { bId: 2, cId: 1 }
  };
  const $router = {
    push: jest.fn()
  };
  const mockStore = new Vuex.Store({
    modules: {
      board: {
        namespaced: true,
        getters: {
          board: jest.fn().mockReturnValue(cardBoard)
        }
      },
      card: {
        namespaced: true,
        actions,
        getters: {
          card: jest.fn(() => fetchedCard.item)
        }
      }
    }
  });
  const options = {
    localVue,
    mocks: {
      $route,
      $router
    }
  };
  test("It should fetch items.", () => {
    wrapper = shallowMount(Card, { ...options, store: mockStore });
    expect(actions.FETCH_CARD.mock.calls).toHaveLength(1);
    expect(actions.FETCH_CARD.mock.calls[0][1]).toEqual({ id: 1 });
  });

  it("has the expected html structure", () => {
    wrapper = shallowMount(Card, { ...options, store: mockStore });
    expect(wrapper.html()).toMatchSnapshot();
  });

  // test("onCLose.", () => {
  //   wrapper = shallowMount(Card, { ...options, store: mockStore });
  //   let button = wrapper.find(".modal-close-btn");
  //   button.trigger("click");
  //   expect($router.push).toHaveBeenCalledWith("/b/2");
  // });
});
