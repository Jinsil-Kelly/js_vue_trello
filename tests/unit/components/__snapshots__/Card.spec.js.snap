// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card has the expected html structure 1`] = `
<modal-stub class="modal-card">
  <div>
    <editcardheader-stub title="card1"></editcardheader-stub>
  </div>
  <div>
    <h3>Description</h3> <textarea cols="30" rows="3" placeholder="Add a more detailed description..." readonly="readonly" class="form-control"></textarea>
  </div>
  <div></div>
</modal-stub>
`;
